### ISSUE TITLE
GHC 9.4.3 installation failure on `ubuntu-20.04`

### ISSUE THREAD

**Issue by andreasabel** [freckle/stack-action#15]
See https://github.com/andreasabel/miniagda/actions/runs/3509343655/jobs/5878377957#step:4:486.
The stack-action fails to install GHC 9.4.3 via stack on `ubuntu-latest`, which is
- Image: ubuntu-20.04
- Version: 20221027.1

The error is:
```
/home/runner/.stack/programs/x86_64-linux/ghc-tinfo6-9.4.3/lib/ghc-9.4.3/bin/ghc-pkg: 
/lib/x86_64-linux-gnu/libc.so.6: 
version `GLIBC_2.32' not found 
(required by /home/runner/.stack/programs/x86_64-linux/ghc-tinfo6-9.4.3/lib/ghc-9.4.3/bin/../lib/x86_64-linux-ghc-9.4.3/libHSrts-1.0.2-ghc9.4.3.so)
```

**Comment by andreasabel** [comment]
As a workaround, I switched to ubuntu-22.04, where this problem does not occur.  However, `ubuntu-latest` is still resolved to `ubuntu-20.04`, thus, this issue does have some relevance.

**Comment by pbrisbin** [comment]
> I switched to ubuntu-22.04, where this problem does not occur ... ubuntu-latest is still resolved to ubuntu-20.04

Do you have this backwards? I thought `ubuntu-latest` was just updated from 20 to 22, and the issue appears on 22, but moving back to 20 is a valid workaround.

I see the following on Jobs using `ubuntu-latest`:

![](https://files.pbrisbin.com/screenshots/screenshot.471187.png)

**Comment by pbrisbin** [comment]
Oh nevermind. This is confusing.

So we are seeing 22.04 being used in _some_ Jobs with `ubuntu-latest`. I'm not sure if this is a bug, or GitHub is starting a transition that's just not evenly distributed. The [official docs](https://github.com/actions/runner-images#available-images) still show `ubuntu-latest` as 20.04.

So, you weren't backwards, apologies. Do I have this right:

| GHC | Ubuntu | Works? |
| --- | --- | --- |
| < 9.4 | `ubuntu-20.04`, `ubuntu-latest` | :heavy_check_mark: |
| 9.4+ | `ubuntu-20.04`, `ubuntu-latest` | :x: |
| 9.4+ | `ubuntu-22.04` | :heavy_check_mark: |

**Comment by andreasabel** [comment]
> Do I have this right

Yes.

But I think whether `ubuntu-latest` resolves to 20.04 or 22.04 is somewhat orthogonal to the issue.  E.g. I do sometimes explicitly ask for `ubuntu-20.04`; thus even if `ubuntu-latest` is bumped, there might be problems for some users.

Since I managed to work around the issue, you could wait for other "customers" to complain before taking action...

A systematic solution could be to set up your "Example" CI so that it tests all the GHA runners (currently it only does `ubuntu-latest`) and all recent GHC versions.

**Comment by pbrisbin** [comment]
Yeah, I agree that's orthogonal if we decide to prioritize a complete fix. That would include all GHCs on all runners Just Work.

But for prioritization purposes, it seems important. I'd consider the GHC of current Stack LTS (9.2) on the runner that is currently GitHub latest (20.04) to be the primary use-case. If that was broken, that'd be very high priority. But if the bug is "to use newer-than-lts GHC you need a newer-than-latest runner", that seems lower priority.

I'm not saying we don't want to make it all work, I just want to make sure I understand the scope of things clearly.

**Comment by andreasabel** [comment]
Personally, I would prioritize this:
> A systematic solution could be to set up your "Example" CI so that it tests all the GHA runners (currently it only does `ubuntu-latest`) and all recent GHC versions.

That would give you an overview of what works and what not.

**Comment by pbrisbin** [comment]
I might be missing something, you're describing (and suggesting we prioritize) how to reproduce and prevent regression on this issue, but not how to fix it? So like, red CI as documentation sort of thing?

**Comment by andreasabel** [comment]
> So like, red CI as documentation sort of thing?

Yeah, unfortunately.  But I guess it is better to have red CI and make existing problems explicit, than having green CI but no coverage...

**Comment by pbrisbin** [comment]
Yeah, that's very fair. Thanks for taking the time to engage on this.

**Comment by pbrisbin** [comment]
I've extended the test suite with a matrix that covers every resolver from nightly back to 12.26 (ghc-8.4) against each OS's `-latest` runner. And I've set our Branch protection to require `*-latest` + `lts` always passes. I think this is the best option because our high priority paths are Green, but other failures may exist as an indication of the state of things. I was pleasantly surprised to see everything besides 20.04 + 9.4 worked, actually.

I think to get `9.4` working on `ubuntu-20.04`, we should look for the [upstream fix](https://gitlab.haskell.org/ghc/ghc/-/issues/22268). A workaround for users today to explicitly choose `ubuntu-22.04` seems acceptable, especially given that that's what `-latest` will be [by March](https://github.com/actions/runner-images/issues/6399).

@andreasabel if you agree, should we close this issue?
